service/src/main/java/uk/gov/hmcts/reform/fpl/service/removeorder/RemoveSentDocumentService.java

```diff
@@ -0,0 +1,110 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static java.lang.String.format;
+import static org.apache.commons.lang3.ObjectUtils.defaultIfNull;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.asDynamicList;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.element;
+import static uk.gov.hmcts.reform.fpl.utils.ElementUtils.unwrapElements;
+
+@Service
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+public class RemoveSentDocumentService {
+
+    public DynamicList buildDynamicList(CaseData caseData) {
+        return buildDynamicList(caseData, null);
+    }
+
+    public DynamicList buildDynamicList(CaseData caseData, UUID selected) {
+        List<Element<SentDocuments>> documentsByParty = defaultIfNull(
+            caseData.getDocumentsSentToParties(), new ArrayList<>());
+
+        documentsByParty.sort(Comparator
+            .comparing((Element<SentDocuments> bundle) -> bundle.getValue().getPartyName()));
+
+        List<Element<SentDocument>> allDocuments = new ArrayList<>();
+        unwrapElements(documentsByParty).stream().forEach(d -> {
+            d.getDocumentsSentToParty().sort(Comparator
+                .comparing((Element<SentDocument> bundle) -> bundle.getValue().getDocument().getFilename()));
+            allDocuments.addAll(d.getDocumentsSentToParty());
+        });
+
+        return asDynamicList(allDocuments, selected, r ->
+            format("%s - %s", r.getPartyName(), format("%s (%s)", r.getDocument().getFilename(), r.getSentAt())));
+    }
+
+
+    public Element<SentDocument> getRemovedSentDocumentById(CaseData caseData, UUID selectedDocId) {
+        return caseData.getDocumentsSentToParties().stream()
+            .map(c -> c.getValue().getDocumentsSentToParty())
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList()).stream().filter(document -> selectedDocId.equals(document.getId()))
+            .findAny()
+            .orElseThrow(() -> new RemovableSentDocumentNotFoundException(selectedDocId));
+    }
+
+    public void populateSentDocumentFields(CaseDetailsMap data, SentDocument sentDocument) {
+        data.put("partyNameToBeRemoved", defaultIfNull(sentDocument.getPartyName(), ""));
+        data.put("sentDocumentToBeRemoved", sentDocument.getDocument());
+        data.put("sentAtToBeRemoved", sentDocument.getSentAt());
+        data.put("letterIdToBeRemoved", sentDocument.getLetterId());
+    }
+
+    public void removeSentDocumentFromCase(CaseData caseData, CaseDetailsMap caseDetailsMap,
+                                           UUID removedDocId) {
+        Element<SentDocument> removedSentDocument = getRemovedSentDocumentById(caseData, removedDocId);
+        final Element<SentDocuments> removedSentDocuments = caseData.getDocumentsSentToParties().stream()
+            .filter(s -> s.getValue().getDocumentsSentToParty().stream().anyMatch(e -> e.equals(removedSentDocument)))
```

Comment:
Could we maybe pull this out into a helper method? Think we do something similar in the above getRemovedSentDocumentById and also again below we do a look up on the documentsSentToParty?

Reply:
Can we maybe reuse removedSentDocuments?